### PR TITLE
feat(eg/subscription): add new resource support

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
add new resource support

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add new resource support.
2. add related document and acceptance test.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/eg' TESTARGS='-run=TestAccEventSubscription_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/eg -v -run=TestAccEventSubscription_basic -timeout 360m -parallel 4
=== RUN   TestAccEventSubscription_basic
=== PAUSE TestAccEventSubscription_basic
=== CONT  TestAccEventSubscription_basic
--- PASS: TestAccEventSubscription_basic (25.55s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/eg        25.622s
```

### CODE REVIEW

File: docs/resources/eg_event_subscription.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+* `targets` - (Required, List) Specifies the list of the event targets.
+  The [targets](#subscription_targets) structure is documented below.
+
+* `description` - (Optional, String) Specifies the description of the event subscription.
+
+<a name="subscription_sources"></a>
+The `sources` block supports:
+
+* `provider_type` - (Required, String) Specifies the provider type of the event source.
```
Comment (deer-hang) [inline docs/resources/eg_event_subscription.md:98]: List available enumeration value types. Check the other parts(line 107, line 132)

File: huaweicloud/services/eg/resource_huaweicloud_eg_event_subscription.go [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+	if err != nil {
+		log.Printf("[ERROR] Invalid type of the %s, not json format", paramName)
+	}
+	return parseResult
+}
+
+func buildEventSourcesOpts(sources []interface{}) []interface{} {
+	result := make([]interface{}, len(sources))
+	for i, val := range sources {
+		source := val.(map[string]interface{})
```
Comment (deer-hang) [inline huaweicloud/services/eg/resource_huaweicloud_eg_event_subscription.go]: It is better to check error when doing type conversion. And check the other parts.
```suggestion
		if source, isMap := val.(map[string]interface{}); isMap {}
```

File: huaweicloud/services/eg/resource_huaweicloud_eg_event_subscription.go [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+	resp, err := subscriptions.Get(client, subscriptionId)
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "EG subscription")
+	}
+	mErr := multierror.Append(nil,
+		d.Set("region", region),
+		d.Set("channel_id", resp.ChannelId),
+		d.Set("name", resp.Name),
+		d.Set("description", resp.Description),
+		d.Set("status", resp.Description),
```
Comment (deer-hang) [inline huaweicloud/services/eg/resource_huaweicloud_eg_event_subscription.go]:
```suggestion
		d.Set("status", resp.Status),
```

File: huaweicloud/services/eg/resource_huaweicloud_eg_event_subscription.go [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+	client, err := cfg.EgV1Client(region)
+	if err != nil {
+		return diag.Errorf("error creating EG v1 client: %s", err)
+	}
+
+	_, err = subscriptions.Update(client, opts)
+	if err != nil {
+		return diag.Errorf("error updating EG subscription (%s): %s", subscriptionId, err)
+	}
+	return nil
```
Comment (deer-hang) [inline huaweicloud/services/eg/resource_huaweicloud_eg_event_subscription.go]: When doing update operation, It is better to run `Read` function to reSet value.
```suggestion
	return resourceEventSubscriptionRead(ctx, d, meta)
```

Comment (Lance52259) [reply huaweicloud/services/eg/resource_huaweicloud_eg_event_subscription.go]: fixed.

File: huaweicloud/services/eg/resource_huaweicloud_eg_event_subscription.go (hunk 3 above)
Comment (Lance52259) [reply huaweicloud/services/eg/resource_huaweicloud_eg_event_subscription.go]: fixed.

File: huaweicloud/services/eg/resource_huaweicloud_eg_event_subscription.go (hunk 2 above)
Comment (Lance52259) [reply huaweicloud/services/eg/resource_huaweicloud_eg_event_subscription.go]: done.

File: docs/resources/eg_event_subscription.md (hunk 1 above)
Comment (Lance52259) [reply docs/resources/eg_event_subscription.md:98]: done.

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/eg' TESTARGS='-run=TestAccEventSubscription_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/eg -v -run=TestAccEventSubscription_basic -timeout 360m -parallel 4
=== RUN   TestAccEventSubscription_basic
=== PAUSE TestAccEventSubscription_basic
=== CONT  TestAccEventSubscription_basic
--- PASS: TestAccEventSubscription_basic (82.94s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/eg        82.999s
```

**Pull request merged**